src/main/java/net/logstash/logback/appender/AbstractLogstashTcpSocketAppender.java

```diff
@@ -531,56 +524,68 @@ public void onEvent(LogEvent<Event> logEvent, long sequence, boolean endOfBatch)
-                     * and the socket was closed by the WriteTimeoutRunnable.
-                     *
-                     * Therefore, attempt reconnection.
-                     */
                     addInfo(peerId + "destination terminated the connection. Reconnecting.");
                     reopenSocket();
-                    try {
-                        readerFuture.get();
-                        sendFailureException = NOT_CONNECTED_EXCEPTION;
-                    } catch (Exception e) {
-                        sendFailureException = e;
-                    }
+                    
+// TODO readerFuture refers to the "old" future and a new one has been started by reopenSocket...
+// Why wait for the old to complete? Is it to reuse its thread and avoid creating/wasting a new one?
+// In this case better to readerFuture.get() before reopenSocket()...
+//
+//                    try {
+//                        readerFuture.get();
+//                        sendFailureException = NOT_CONNECTED_EXCEPTION;
+//                    } catch (Exception e) {
+//                        sendFailureException = e;
+//                    }
                     continue;
                 }
+                
+                /*
+                 * Write the event in the output stream.
+                 * Drop event if encoder throws an exception.
+                 * Reconnect if an exception is thrown by the connection stream itself.
+                 */
                 try {
                     writeEvent(socket, outputStream, logEvent, endOfBatch);
                     return;
+                    
+                } catch (EncoderException e) {
+                    /*
+                     * Encoding threw an exception. Warn and drop event before it becomes a "poison".
+                     */
+                    addWarn(peerId + "Encoder failed to encode event: " + e.getMessage() + ". Dropping event.", e);
+                    fireEventSendFailure(logEvent.event, e);
+                    break;
+                    
                 } catch (Exception e) {
-                    sendFailureException = e;
-                    addWarn(peerId + "unable to send event: " + e.getMessage() + " Reconnecting.", e);
                     /*
-                     * Need to re-open the socket in case of IOExceptions.
-                     *
-                     * Reopening the socket probably won't help other exceptions
-                     * (like NullPointerExceptions),
-                     * but we're doing so anyway, just in case.
+                     * Any other exception is thrown by the socket stream (or bug in the code).
+                     * Re-open the socket and get a fresh new stream.
                      */
+                    addWarn(peerId + "unable to send event: " + e.getMessage() + " Reconnecting.", e);
                     reopenSocket();
                 }
```

Comment:
Correct. It repeats until one of the following condition applies:
- the event is successfully sent
- the event cannot be encoded in which case a warn status is emitted and the event is dropped (poison)
- the appender is stopped

Any IOException received while writing into the stream or attempting to connect to the remote destination causes a reconnect. Dropping the event in this case and restarting the loop with the next one doesn't make much sense here.

The idea behind these changes is an event should be dropped only if it cannot be encoded. Destinations issues should not be a cause. If needed, we can introduce the concept of black listing of miss-behaving destinations to remove them from the pool...

Reply:
In any of these conditions, will this get into a busy loop (consuming too much CPU)?  i.e. does this need to back-off on retry?

Reply:
No more risk than before... It loops until the event is successfully sent or encoding fails in which case the event is dropped. Basically, the only reason why it may loop is because of a network issue. 
The previous implementation looped only 5 times then discarded the event and restart with the next event of the ring buffer. Since the issue is IO-related, it is likely to loop 5 times again, drop and so on... ultimately, it drains the ring buffer (without notice) until the IO problem is resolved.
This new implementation behaves the same except it does not drop the event but loops with the same event until the IO problem is gone.

Furthermore, `openSocket()` already implements a back-off strategy that kicks-in if the last attempt to connect to the selected destination is too "close" (prevent reconnection storms).